New interfaces & fix Outlet jacobiany for several domains

hwpang opened the pull request:
1. For the first two commits, Condensation/evaporation type interfaces are added and are intended for liquid phase domains.
2. For the fourth commit, a bug related to the Outlet part of the jacobiany were fixed. Non-gas-phase domains' N is calculated via `sum(ns)` as opposed to `PV/RT` in gas-phase domains. The mass flow from Outlet is `-flow*ns/N`, so the derivative d/dni (-flow * nj/\sum_j nj) should be -flow/N + ni/N^2 when j == i and nj/N^2 when j != i.
3. The last commit fixes a typo related to the Outlet jacobiany for ConstantTPDomain, i.e. d/dV (-flow * nj/N) = d/dV (-flow * nj * R * T/P/V) = flow * nj * R * T/P/V^2 = flow/N/V.

hwpang commented:
@mjohnson541 As per offline discussion, these are the equations used for the vapor-liquid mass transfer. The contribution from the vapor-liquid mass transfer is

`dy/dt = reaction terms + flow terms + kLA/kH (P_vap*x_vap/(R*T_vap)-cs)*V`, where the condensation contribution is `kLA/kH P_vap*x_vap/(R*T_vap)*V` and the evaporation contribution is `-kLA/kH cs * V`, where kLA, kH, x_vap, cs are vectors for each species

hwpang commented:
The new commit here removes the typo related to kH, which shouldn't appear in the evaporation contributions.

mjohnson541 commented on src/Domain.jl:
```diff
@@ -1872,6 +1872,7 @@ end
             flow = inter.F(t)
             @simd for i in domain.indexes[1]:domain.indexes[2]
                 @inbounds @fastmath jac[i,i] -= flow/N
+                @inbounds @fastmath jac[i,:] .+= flow*ns[i]/(N*N)
```
This looks right physically and mathematically even if it ruins sparsity XD, but from a simulation standpoint it's kind of terrible, because our most common case involving this is a liquid reactor with an outlet and normally your liquid isn't going to expand to the size of the container it will just effectively be a smaller volume. Not that this needs fixed in this PR, but we should probably think about how we handle an outlet on a liquid reactor because it'd be super easy for someone to setup a liquid reactor simulation with an outlet like this and get a weird result.

mjohnson541 commented on src/Domain.jl:
```diff
@@ -1427,6 +1427,10 @@ export calcthermo
             dydt[d.indexes[1]:d.indexes[2]] .+= inter.y.*inter.F(t)
         elseif isa(inter,Outlet) && d == inter.domain
             dydt[d.indexes[1]:d.indexes[2]] .-= inter.F(t).*ns./N
+        elseif isa(inter,Condensation) && d == inter.domain
```
Can I model kinetics in a gas that's over a constant composition liquid with these? I think you have all the parameters for that in these, but this configuration here I think only works if the domain is a liquid domain? 

Do we have a failsafe that warns the user if they try to use one of these new interfaces and it doesn't work with the domain they chose?

hwpang replied:
That's true, I can generalize these interfaces such that it could work for a vapor domain that's over a constant composition liquid. I will add the failsafe warnings. Thanks!

hwpang replied:
I generalize the interface to gas phase domains and add failsafe warning to the interface

mjohnson541 reviewed (commented):
Okay, I think this PR brings up some questions about how we do interfaces that we didn't think about before.

hwpang commented on src/Domain.jl:
```diff
@@ -1634,7 +1634,7 @@ end
                 @inbounds @fastmath jac[i,i] -= flow/N
             end
-            @views @inbounds @fastmath jac[domain.indexes[1]:domain.indexes[2],domain.indexes[3]] .-= flow.*ns.*(-C)
+            @views @inbounds @fastmath jac[domain.indexes[1]:domain.indexes[2],domain.indexes[3]] .-= -flow.*ns/(N*V)
```
This line is calculating the Outlet contribution to the jacobiany d/dV dni/dt, where the dni/dt from the Outlet is flow*ns/N, which is flow*ns/(P*V) *RT, and d/dV flow*ns/(P*V) *RT = flow * ns * (-1/V^2) * 1/P * RT = flow* ns/N *(-1/V)

hwpang commented on src/Interface.jl:
```diff
@@ -243,3 +243,33 @@ struct Outlet{V} <: AbstractBoundaryInterface
 export Outlet
+
+
+struct Condensation{S,V<:AbstractArray,U<:Real,X<:Real} <: AbstractBoundaryInterface
```
Yes this is the one-way condensation because the evaporation direction depends on the liquid composition. I'm thinking to change the name `condensation` to `VaportoLiquidMassTransfer`, and `evaporation` to `LiquidtoVaporMassTransfer`. What do you think?

hwpang replied:
Evaporation and condensation are now combined to one interface

hwpang commented on src/Interface.jl:
```diff
@@ -243,3 +243,33 @@ struct Outlet{V} <: AbstractBoundaryInterface
 end
 export Outlet
+
+
```
Thanks! I will add the docstrings!

hwpang replied:
I change the name of the interface to kLAkHCondensationEvaporation to reflect that we use kLA and kH to model evaporation and condensation. Docstring is added too

hwpang commented on src/Interface.jl:
```diff
@@ -243,3 +243,33 @@ struct Outlet{V} <: AbstractBoundaryInterface
+    x_vap::V
+    T_vap::U
+    P_vap::X
+    kLA::Array{Float64,1}
```
kLA and kH are now appended to Species object

mjohnson541 commented:
It looks like the sensitivities and fluxes in that model are changing as a result of this PR resulting in a different analysis string. If you don't think you've changed any code that affects it that probably means there's a bug in the PR. If the change is expected you can swap out the current test string for the one in the test error message.

hwpang commented:
I ran the test on main branch (`4949951ec60659209b09f1b6c23734760e772a4f`) locally and encountered the same error. I also ran the test on commit `93a675363b9bc529b83b5581c329edcaabf923fd` where this test was added, and it failed with the same error. Both these two and this PR results in the same evaluated string in the error message. So I think it's safe to update the reaction analysis string.

hwpang commented:
@mjohnson541 The newest push changes the reaction analysis string and add new interface `ConstantVolumetricFlowRateOutlet`

mjohnson541 reviewed (commented):
This looks mostly good to go. I think it'd just be good to add comments on the parts that are the result of more complex math. We haven't really done it in the past, but we have these lines so many places it's probably best to include the math in a comment next to the code at least for the ones that aren't super simple.

mjohnson541 commented:
Also can you remove the underscore in the V_out field name? I don't think we do that with any of the other variable names and it's also annoying to type. Also I think you need to rebase onto main rather than master.

hwpang commented on src/Domain.jl:
```diff
@@ -1857,6 +1862,15 @@ end
+                @fastmath dTdt = (P*inter.V_out(t))/(N*Cvave)
+                @fastmath ddnidTdt = -dTdt*(dCvavedni/Cvave)
+                @inbounds jac[domain.indexes[3],i] -= ddnidTdt
+                @inbounds @fastmath jac[domain.indexes[4],i] -= P/T*ddnidTdt 
```
I redo the math for all interfaces and debugged the jacobians and added comments. It should be much clear now!

hwpang commented:
@mjohnson541 I think the new push addresses all the comments! I also loosened the type constraint for interface conddict to make using interfaces easier from pyrms.

hwpang commented:
I changed a small typo in equation

hwpang commented:
Removed typo

hwpang commented on src/Interface.jl:
```diff
@@ -252,3 +252,52 @@ struct Outlet{V,FF<:Function} <: AbstractBoundaryInterface
+
+struct kLAkHCondensationEvaporation{S,V1<:AbstractArray,V2<:Real,V3<:Real,V4<:AbstractArray,V5<:Real,V6<:AbstractArray,V7<:AbstractArray} <: AbstractBoundaryInterface
+    domain::S
+    x::V1
```
During developer hour: change x to molfraction

mjohnson541 commented:
Apart from the mole fraction thing and either splitting it apart or having some extra comments in the docstring about the interface being for either liquid or gas reservoir I think this is good.

hwpang commented:
@mjohnson541 New push change the interface names, the variable names,  and add a more detailed docstring as we discussed offline!

mjohnson541 commented:
Looks good can you squash?

hwpang commented:
@mjohnson541 The new push squashed the commits, and updated the commit message to reflect the change in interfaces' names

hwpang commented:
New push rebased main

mjohnson541 commented:
Sorry, one last thing can you update master -> main in the workflow files so the CI runs?

hwpang commented:
@mjohnson541 The only tests failed is due to the file `../src/testing/superminimal_broken.rms` was never added in the original PR #174. Back then we haven't changed the CI workflow to be triggered by the main branch PR, so this bug wasn't caught. Could you upload the file?

hwpang commented:
New push removes bugs in ConstantTPDomain interface jacobiany

hwpang commented:
New push removes bugs in ParametrizedTPDomain interface jacobiany

hwpang commented:
@mjohnson541 All tests have passed. Could you review this PR? Thanks
